Share2.0 sharemanager

rullzer opened the pull request:
@schiesbn @DeepDiver1975 first draft is here.

for now we have a sharemanager object and a ishareprovider and that is it. So no share object. Basically everything goes trough the share manager.

The manager itself acts like some kind of middleware between the shareprovider and the OCS Share API. Making sure that arguments are in the right format (date for example). And making sure that the user we want to share with exists etc.

IShareProviders register themself at the sharemanager with a callback function. And then all the magic happens.

Many things in the code do not work as expected yet or are even plain wrong now. But if we agree this is the way to go I can start cleaning up and already write some simple unit tests for this part. And do some squashing of commits :P

So feedback please :)

DeepDiver1975 commented:
@rullzer 

```PHP Fatal error:  Interface function OC\Share20\IShareProvider::share() cannot contain body in /ssd/jenkins/workspace/core-ci-linux@2/database/sqlite/label/SLAVE/lib/private/share20/ishareprovider.php on line 25

DeepDiver1975 commented:
@rullzer seriously get a proper IDE - and this is no recommendation - but an order :wink: 

![bildschirmfoto von 2015-10-19 22-09-43](https://cloud.githubusercontent.com/assets/1005065/10589903/34f20efa-76ae-11e5-92bc-472bf1cc58b3.png)

rullzer commented:
@DeepDiver1975 yeah I know it doesn't work. But I first wanted to agree if this was the way to go for the manager before I start to do all kinds of actual code and add unit tests etc.

DeepDiver1975 commented on lib/private/share20/exceptions/sharenotfoundexception.php:
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace OC\Share20\Exceptions;
```
I'm not sure about the namespace Share20 ... we will find something better

rullzer replied:
Agreed. For not is just there since `OC\Share` was taken

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	 * @param \DateTime $expireDate
+	 * @param $password
+	 */
+	public function share(\OCP\Files\Node $path,
```
I still miss CRUDS operations as discussed

- createShare
- updateShare

schiesbn replied:
yes, this method should be called "createShare" or just "create"

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	 * @param $password
+	 */
+	public function share(\OCP\Files\Node $path,
+	                      $shareType,
```
And in terms of extensibility we might consider passing in an assoc array

nickvergessen replied:
or an object?

nickvergessen replied:
I hate magic assoc arrays

rullzer replied:
So what kind of extentions are you thinking about? As far as I know this list of arguments has been stable for a long time.

I'm not a fan of the big magic assoc array as well.

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	/**
+	 * Get all shares by the given user
+	 *
+	 * @param IUser $user
```
return missing

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	 * @param int $id
+	 * @param \DateTime $expireDate
+	 */
+	public function setShareExpirationDate($id, \DateTime $expireDate);
```
this has to be done in updateShares

rullzer replied:
So 1 big updateShares function? Then we get again 1 big function. Where stuff is passed and we need to first figure out what actually changed. And then if conflicting info has changed we need to figure that out... throw errors etc that have to return enough info to the user that they know what went wrong.

rullzer replied:
So what function signature are you thinking of?

`updateShare($id, $key, $value)`

or

`updateShare($share)`

?

First one is fine in my opinion. Second one, not so much.

DeepDiver1975 replied:
CRUD operations ... RESTfull api design .. yes: one update method

schiesbn replied:
I agree, I would also prefer something like ````updateShare($id, $key, $value)````

nickvergessen replied:
The disadvantage of `$key` is that you have a string, which can have typos and stuff...

schiesbn replied:
I discussed this with @nickvergessen and I tend to agree. A object has the advantage that it has well defined and documented get/set methods and no such things like typos in the string which defines the key, etc.

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	 * @param int $id
+	 * @param int $permissions
+	 */
+	public function setSharePermissions($id, $permissions);
```
this has to be done in updateShares

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	 * @param int $id
+	 * @param string $password
+	 */
+	public function setSharePassword($id, $password);
```
this has to be done in updateShares

DeepDiver1975 commented on lib/private/share20/ishareprovider.php:
```diff
@@ -0,0 +1,115 @@
+	 * @param int $id
+	 * @param string $password
+	 */
+	public function verifySharePassword($id, $password);
```
what is that used for?

rullzer replied:
To verify is a link share has a valid password. When the password is verified the storagae can directly update the hash if that is required. We are missing this currently so on long running instances we can't migrate to our new and improved hashing.

DeepDiver1975 replied:
but password verification is something to be done within the manager or on create/update in the share provider

rullzer replied:
Passwords of public link shares need to be verified.

rullzer replied:
So it is not checking if the password is strong enough etc. But rather if it is right.

schiesbn replied:
Just thinking, couldn't this be part of the getShare method? In case of a public link getShare() would verify the password and only return the share if the password is valid?

rullzer replied:
`getShareByToken` you mean... yeah that would be a good idea

DeepDiver1975 commented on lib/private/share20/manager.php:
```diff
@@ -0,0 +1,421 @@
+	 * @param int[] $shareTypes
+	 * @param callacble $callBack
+	 */
+	public function registerShareProvider($id, $displayName, $shareTypes, callable $callBack) {
```
we dont need a register operation - as discussed the manager will get the providers from the home storage implementation

schiesbn replied:
what about the other share providers like the public link share provider? Probably it would be still good to have a method to register additional share providers.

schiesbn replied:
@DeepDiver1975 after discussing it with @rullzer I updated the main sharing 2.0 issue with more details how the share provider works and why we need both, a way to register a share provider and the ability that the manager ask the primary storage for a share provider.

Please look here under the headline "shareprovider": https://github.com/owncloud/core/issues/19331

DeepDiver1975 commented on lib/private/share20/manager.php:
```diff
@@ -0,0 +1,421 @@
+	/**
+	 * Retrieve all share by the current user
+	 */
+	public function getShares() {
```
we need a pagination mechanism - think big - we don't want to merge two arrays of 500000 elements

DeepDiver1975 replied:
We should consider to use some iterator pattern maybe

rullzer replied:
I agree that we would want pagination here but then that would result in strict ordering. We can force the IShareProviders to have strict ordering (and do proper pagination). But here we then would require first the local and then the federated shares for pagination for example.

rullzer replied:
Also I can imagine certain providers not providing this functionlity since it might simple be to expensive (for example if you store it on disk with the file metadata).

DeepDiver1975 replied:
> Also I can imagine certain providers not providing this functionlity since it might simple be to expensive (for example if you store it on disk with the file metadata).

well - any system must have a mechaism to deliver a list in chunks - otherwise this will not work.

rullzer replied:
yeah agreed. We don't care how the provider does it... as longs as they do it.

rullzer commented:
So as discussed with @schiesbn and @nickvergessen there is now a share object which is passed around so we keep the general logic in 1 place. This looks a lot cleaner than it was before. It keeps a lot of logic at the right place.

This means that the manager and providers now have `create` `update` and `delete` functions. That just take a share object and do their thing with it.

And that is enough share manager for today.

DeepDiver1975 commented on lib/private/share20/manager.php:
```diff
@@ -0,0 +1,273 @@
+
+		//TODO: ID's should be unique who handles this?
+
+		$shares = array_merge($storageShares, $federatedShares);
```
since merging these results has it's drawbacks we might want to consider getShares to be specific for the share type

nickvergessen replied:
yeah, the "TODO proper pagination" will teach you that merging arrays will not work together with paging, remember the sharee mess.

rullzer replied:
@nickvergessen that was my point exactly. Doing it properly means we either have to define an order ourself. i.e.  user, group, link, remote.

Since I think just limiting them to the sharetype is not really what people expect (and will also break the current OCS API behaviour)

DeepDiver1975 replied:
> Since I think just limiting them to the sharetype is not really what people expect (and will also break the current OCS API behaviour)

What is the existing behavior with respect to OCS API? Can it properly perform pagination?

rullzer replied:
> What is the existing behavior with respect to OCS API? Can it properly perform pagination?

No selection on share type. No pagination just everything is returned.

DeepDiver1975 replied:
> No pagination just everything is returned.

okay - so this has to change soon-ish - after sharing 2.0 - not too many changes at a time

DeepDiver1975 replied:
OCS API is returning local shares as well as fed shares?

rullzer replied:
> OCS API is returning local shares as well as fed shares?

Yes and no... the remote shares initiated by the user are in there. But the shares received by the user are currently not shown in shared with me (because I don't know how to properly fix this in the old code)

nickvergessen replied:
They can't (and shouldn't) be returned in one list, because the id (unique autoincrement from db) belongs to two different tables

DeepDiver1975 replied:
> because the id (unique autoincrement from db) belongs to two different tables

the idea was to prefix these ids with a provider prefix ... but yes this design is kind of questinoable under the light of pagination

rullzer replied:
They the getShareById function should also require an shareType. Which means that routes like

`shares/<ID>` are no longer valid. 

And they are now.

DeepDiver1975 replied:
> shares/<ID> are no longer valid.

What ID can that be? Fed shares as well? I don't think so - right?

rullzer replied:
that can currently be any initiated sharing id. So if I share to you a federated share... and that on my side has id 5 I can get it via `shares/5`

rullzer replied:
Updated the code with how I think we can do proper pagination on this.

rullzer commented:
I added the register functions back as dicussed with @schiesbn. Since we want our own implentations to be able to register themselfs so they can be used as fallback if the storage does not provide it.

Also I modified the manager a bit to allow for proper pagination of shares by ordering them by shareType.

DeepDiver1975 commented:
> I added the register functions back as dicussed with @schiesbn. Since we want our own implentations to be able to register themselfs so they can be used as fallback if the storage does not provide it.

whatfor? the default implementation is provided by the default implementation of the home storage.
there is no need to provide any provider any other way.

rullzer commented:
> whatfor?

With all the effort we spend moving core functionality to apps this seems logical.

karlitschek commented:
@rullzer Which effort to move core functionality to apps do you mean?

rullzer commented:
> @rullzer Which effort to move core functionality to apps do you mean? 

Well we move webdav to a separate app. There are efforts to get the preview generation out of core and only keep the manager etc.

karlitschek commented:
I'm not aware of any work to move previews out. So this is not a general theme or indication that we should do this ll over the place :-)

schiesbn commented:
> whatfor? the default implementation is provided by the default implementation of the home storage.
there is no need to provide any provider any other way.

we still have the "federated share provider" which will be a app and you don't want that the app manager knows the app name, need to check if the app is available and then call it. The other way around makes just much more sense.

Also I'm not sure yet if the default share provider for user, group and link share should be implemented as part of our storage because we also need it as fall back if a different storage is used but didn't implement a share provider. Therefore it makes much more sense if we have a "federated share provider" and a "oc default internal share provider" both implemented as app and which register them self at the share manager. The share manager can additionally ask the storage "getShareProvider()" if something is returned fine, then we use it. But it can also happen that the storage will just return false because the storage doesn't provide his own share provider, in this case we will fall back to the "oc default internal share provider"

DeepDiver1975 commented:
> I'm not aware of any work to move previews out. So this is not a general theme or indication that we should do this ll over the place :-)

@karlitschek with 8.2 we started to allow additional preview generators to be added by apps using public api. So it's possible to have new preview generators in apps and even move e.g. the libreoffice based preview generators to aps on their own which is encouraged by @LukasReschke if I recann correctly due to potential vulnerabilities.

DeepDiver1975 commented:
@schiesbn point taken regarding the fed sharing - given the all the discussions above I don't yet finally know where this all fits together ... let's see.

regarding the homestorage based share provider:

```php

interface IStorage {
  function getSharePoriver();
}

class Common implemets IStorage {
  function getShareProvider() {
    return new DefaultImplementation();
  }
}

class ObjectStore extends Common {
  function getShareProvider() {
    return new MyVeryOwnImplementation();
  }
}

class StorageWhichUsesDefault extends Common {
  // simply don't implement the method
}

class ShareManager {
  function getShareProvider($path) {
    $this->getStorageForPath()->getShareProvider();
  }
}

rullzer commented:
@DeepDiver1975 
```php
class ShareManager {
  function getShareProvider($path) {
    $this->getStorageForPath()->getShareProvider();
  }
}
```

Just to clarify, do we then expect the storage to take care of all the magic with finding the right provider? E.g. You have a file on external storage, which does not handle its own sharing. Will it then still give us the primary storage implementation?

And if the primary storage does not provide an implementation does it then automagically fetch our default implementation.

I like the abstraction here, and it makes our share manager a lot simpler. But will most likely make the storage backend a little bit more complicated.

DeepDiver1975 commented:
> Just to clarify, do we then expect the storage to take care of all the magic with finding the right provider? E.g. You have a file on external storage, which does not handle its own sharing. Will it then still give us the primary storage implementation?

as for now we anyhow only support the homestorage/primary storage to specify the one and only share provider. as for now we do not support additional providers on e.g. external storages - which could be an interesting thing in the future

nickvergessen commented on lib/private/share20/share.php:
```diff
@@ -0,0 +1,252 @@
+	 * @return Share The modified object
+	 */
+	public function setInternalId($id) {
+		$this->id = $id;
```
missing `return $this;`

nickvergessen commented on lib/private/share20/share.php:
```diff
@@ -0,0 +1,252 @@
+								$sharedBy = null,
+								$shareOwner = null,
+								$internalId = null,
+								$providerId = null) {
```
I'd like to not have those here. Just make people call the setters manually afterwards?

rullzer replied:
Was on my mind as well. So just the path, shareType, ShareWith and permission then. Since that is at least required for a share.

nickvergessen replied:
Or nothing at all? That's what we do on the notifications:
https://github.com/owncloud/core/blob/master/lib/private/notification/notification.php#L71-L74

This allows you to set only a shareType and then you can delete all shares of a given type with one call.
Othewise you have to create thousands of objects to do so.

DeepDiver1975 replied:
I personal prefer to either have the data set in the ctor or via setters - depending on the usecases.

But seriously this is something to discuss later - let's first get the interfaces properly sorted out

rullzer replied:
>  This allows you to set only a shareType and then you can delete all shares of a given type with one call.
Othewise you have to create thousands of objects to do so.

I like this. Let me add setters for all fields.

nickvergessen commented on lib/private/share20/manager.php:
```diff
@@ -0,0 +1,324 @@
+	 *  'groups' => ['group2']
+	 * ]
+	 *
+	 * This is required for encryption
```
and activities and others

rullzer replied:
@nickvergessen do you need this exact function or something more general. Because I was thinking about a general traversal function. Give it a path and it will give you back (and iterator maybe?) with all the shares going up to the root.

nickvergessen replied:
I need a method which I give a path and get back a list of users with their respective path for that very same file.
Currently I use `getUsersSharingFile()` for that with the 3rd and 4th argument set.

nickvergessen replied:
Although it's not working completly correct atm https://github.com/owncloud/core/issues/15858

rullzer commented:
So as I see it now we are kind of converging to a setup where a `Share` object just holds data. We will do no active checks on this data during the getting and setting phase.

When we tell the share manager do something (create, update, delete) we pass it a share object.
In the share manager we can do basic sanity checks. For example is a user passed if the sharetype is USER etc. We can also do a little bit more generic checks such as if we enforce an expiration date and non is set set it, or if it is in the future throw an exception etc.

Now @nickvergessen pointed out that a nice feature could be that if you pass `deleteShare` a `Share` object where only a path is set then you can interpet this as delete all the shares for this path. Same holds for the shareWith field etc. I think this would make sense.
I think it can even make sense for `updateShare` but then we need to properly define behaviour.

DeepDiver1975 commented:
> Now @nickvergessen pointed out that a nice feature could be that if you pass deleteShare a Share object where only a path is set then you can interpet this as delete all the shares for this path. Same holds for the shareWith field etc. I think this would make sense.
> I think it can even make sense for updateShare but then we need to properly define behaviour.

exactly considering such magic convenience functinoalities did lead to the current implementation mess :-1: 

keep it simple!

delete is for one object identified by the id
update is to update exactly one object

karlitschek commented:
@DeepDiver1975 Yes I know. But this doesn't necessarily mean that we MOVE more previewers out of core. :-)

LukasReschke commented:
> @DeepDiver1975 Yes I know. But this doesn't necessarily mean that we MOVE more previewers out of core. :-)

What would make sense is to have an application `previewgenerator`, living in core, in the apps folder, being shipped and enabled by default. This app registers all preview providers that we have enabled by default. So if you don't want to have previews you can just disable the app.

The already existing, by default disabled preview generators, should however not live in core. If somebody really wants a preview generator for stuff like "Fonts" or "Office files". They can for example just install an app "Office Preview" from the AppStore and *tadaa*, you get nice preview of office files. Of course, we should provide these as an app that then the community can maintain. We don't really should do it.

This gives more flexibility to developers (as we already denied quite some preview generators in core due to security concerns) as well as keeps the existing default one in core and makes core smaller. We already have the required code to allow apps to register preview providers and it would, from a security PoV, perfectly make sense.

Less code paths and all the risky stuff not being shipped with server. It's already disabled and really increases the risk.

nickvergessen commented:
Yo, this is about Share Manager, can you please discuss moving stuff to other apps in other issues please?!
Sorry if this sounds rude...

karlitschek commented:
@nickvergessen Valid point. Sorry. By the way: Have you seen the new Start Wars trailer? ;-)

rullzer commented:
> keep it simple!
> 
> delete is for one object identified by the id
> update is to update exactly one object

Ok agreed. The general use case is that anyway since the external API is the OCS Share API which only allows single deletion/modification anyway.

rullzer commented:
@DeepDiver1975 
> as for now we anyhow only support the homestorage/primary storage to specify the one and only share provider. as for now we do not support additional providers on e.g. external storages - which could be an interesting thing in the future

I get that. But what I mean is who is responsible for getting the right share provider? The manager? So will the manager ask the primary storage (need to figure out if we can inject that) give me your share provider. This either succeeds or not. And if not it will ask the home storage (also need to inject that) for the default implentation.

Or do we do it the other way around. We have our path which is a Node object. We ask that node object give me your storage and then ask that give me your shareprovider. If that node does not implement it will just throw it to its parent and so on.

Like I said who do we want to hand this off to? I think the second approach is cleaner. But I have no insight how much work it would be to do that on the FS level instead of in the share manager.

rullzer commented:
Besides the stuff where to get the shareprovider does anybody have major further comments? Else I'll start with writing some unit tests and then making sure they pass for the basic sanitiy checks in the manager.

schiesbn commented:
@rullzer I agree, let's start implementing it. I think we have a good base now to move forward. We can still adjust stuff if we continue to integrate the existing APIs and develop the share providers if we discover that some stuff need to be added/changed.

DeepDiver1975 commented:
So .. let's move on to implement the first pass through for one share type - I suggest to use the token based or the user .. I don't care much.

Pass through for me means:
OCS -> Sharemanager -> shareprovider -> existing database table

Regarding the database access: I want one class to handle the sql statements only.

DeepDiver1975 commented:
> I get that. But what I mean is who is responsible for getting the right share provider?

The storage. Only the storage knows its sharing implementation.

schiesbn commented:
> The storage. Only the storage knows its sharing implementation.

That's what the manager should know. The share provider register themself (or get fetched by the manager) for different share types. If then a share of type X come in the manager know that he has to pass it on to share provider Y

rullzer commented:
Ok next round. I started with a single pass. Lots of stuff needs to be added sorted out etc. But as said many times lets get things in in little steps so we don't end up with a 100K lines PR.

Right now there is a new OCS Share API endpoint `../v2/shares`. I propose to make that the new furute endpoint. This means we are stricter here than we are on the current `../v1/shares` endpoint. For example there only ISO8601 dates ('Y-m-d') are accepted and returned.

For now only creating a share works. And we are missing checks and tests for that still.

I added a new table `share_20` this is just a WIP table. In the end we will morph the new table to the final table we decide on. I did add a unique constraint on the table. This means that the database actually keeps track if an identical share already exists. Saves us a query and makes the DB do what it should.

Please have a look :)

DeepDiver1975 commented:
> Right now there is a new OCS Share API endpoint ../v2/shares. I propose to make that the new furute endpoint. This means we are stricter here than we are on the current ../v1/shares endpoint. For example there only ISO8601 dates ('Y-m-d') are accepted and returned.

no - we should not introduce a new OCS API - we do not want to rebuild everything

DeepDiver1975 commented on lib/private/server.php:
```diff
@@ -521,6 +521,17 @@ public function __construct($webRoot) {
 				$request
 			);
 		});
+		$this->registerService('ShareManager', function(Server $c) {
```
I see no need in registering the ShareManager until we finally did agree on the OCP

rullzer replied:
Fair enough.

rullzer commented:
> no - we should not introduce a new OCS API - we do not want to rebuild everything

We can replace it later. But for now this is way easier than trying to get 2 code paths in 1 endpoint.

DeepDiver1975 commented:
> We can replace it later. But for now this is way easier than trying to get 2 code paths in 1 endpoint.

I'm willing to accept this as a temporal solution - please don't name is v2 then :wink:

rullzer commented:
> I'm willing to accept this as a temporal solution - please don't name is v2 then :wink:

Fair enough. It is now `api/temp/shares`

rullzer commented:
@schiesbn @DeepDiver1975 should we get this in? I'd really like not to make the PR any bigger. At least something is working now. So then we can continue to add stuff in smaller PR's.

DeepDiver1975 commented:
> @schiesbn @DeepDiver1975 should we get this in?

hmmmm .... I honestly uncertain because of:
- temp route on the API
- the new table

Let me think about this ....

@schiesbn your opinion? THX

DeepDiver1975 commented:
@rullzer please rebase and run the sharing integration tests (and now the fun starts because of the new ocs api route)

DeepDiver1975 commented:
Okay - we have to kill the temp route and integrate the new code with v1 ocs api - anything else is pointless

rullzer commented:
@DeepDiver1975 would a config variable (in config.php or something) that sets useSharing2 be an OK solution?

rullzer commented:
So the new table is just temporary. But else you can't switch between the two implementations. I really think we need it for now

rullzer commented:
> @DeepDiver1975 would a config variable (in config.php or something) that sets useSharing2 be an OK solution?

But that would still require two separate jobs to be run. So then you could also let one of the jobs run against the new endpoint.

I get that it would be very useful to intergrate the new code directly into the existing endpoint. But I don't think this is something that can happen from the start. For example we will make some changes to the DB (don't have to be huge but they will be made). I don't think we want rewrite the current sharing code to still work with those changes. 

From my POV it is way easier to just have the 2 endpoints. And once the new endpoint is in a state that all tests nicely pass we can migrate. Then we can have 1 point in time where we do a migration step. Fix all the stuff that might be broken in the old table. Add new columns, drop the ones we don't need. Etc etc.

schiesbn commented:
> So the new table is just temporary. But else you can't switch between the two implementations. I really think we need it for now

yes, I would add a config.php switch (default should the the old sharing). This allows us to switch between both implementation until sharing 2.0 is finished and ready for productive use.

DeepDiver1975 commented:
But honestly: I don't want that temp stuff ever to hit master - never

rullzer commented:
> But honestly: I don't want that temp stuff ever to hit master - never

I share that concern. How about this

We create a branch sharing2.0 which we daily rebase on master. There should not be much overlap but still.
That way we can break the endpoint all we want. And keep working until we have passing tests. And then once everything is ready we merge sharing2.0 back into master.

We do PR's of the stuff we do against this sharing2.0 branch.

That way we still do everything in small incremental steps. But keep it separated until it works.

The downside is of course one big PR at the end. But since we rebase daily and have reviews of the PR's this should not less of a problem.

rullzer committed:
Initial setup of shareing 2.0 sharemanager and share

* Added sharemanager class

This is the central class where all API calls talk to (OCS/Activity).
This in turn talks to the share providers to get the actual sharing
done.

It uses all ShareObjects

* Added share class

Simple class to hold all the share properties that is passed around

* Added IShareProvider interface

Interface that providers os a share have to implement.

rullzer committed:
Prepare OCS Share endpoint

rullzer committed:
The new sharing code now handles deletion

OCS -> ShareManager -> DefaultShareProvider

rullzer commented:
Ok branched of the work I had on createShare. (Don't worry it is not deleted). And chose to work with share deletion instead. Since share creation is much much more complex. And I want to have something in. 

*IF* this is merged OCS will use the delete of the share manager.

rullzer commented:
TODO:
- [ ] Add hooks to code
- [ ] Check activities

rullzer commented:
Ok so some issues with this. Lets decide if they are showstoppers or not.

- Emit hooks: Do we want this/need this? Can we fix it later? I assume activities is depending on the hooks @nickvergessen ?

- The old sharing code updated the parent of child shares on deletion. This code does not. I know this is most likely an edge case that is not used that much. But is it acceptable for now?

- Federated delete is not handled properly yet. As in no unshare stuff is send.

rullzer commented:
I suggest to modify the OCS code to still take the old code path for federated shares. Then at least that keeps working.

As for the complex unshare with parent stuff. Probabaly fine to keep it as it is. As we want to move do a flat model anyway.

nickvergessen commented:
> Emit hooks: Do we want this/need this? Can we fix it later? I assume activities is depending on the hooks @nickvergessen ?

Would be nice if they continue to work. But if they don't work now, people will have to use other activities to test the system.

> The old sharing code updated the parent of child shares on deletion. This code does not. I know this is most likely an edge case that is not used that much. But is it acceptable for now?

This will cause problems with syncs and mostlikly fail Smashbox, unless I'm misunderstanding it?

rullzer commented:
@nickvergessen could you run the sharing smashbox tests against this PR and see what fails?

DeepDiver1975 commented:
for activities we should move away from hooks to an explicit implementation of a activities provider

DeepDiver1975 commented:
but other code elements still rely on the hooks

rullzer commented:
Let me add back the hooks. It "should" not be that much work

rullzer committed:
Use the old code for remote shares for now

rullzer committed:
Add unshare hooks

rullzer commented:
Ok remote share deletion is now still handled by the old code.
And hooks are back!

rullzer commented:
So what we don't do now is in https://github.com/owncloud/core/blob/master/lib/private/share/helper.php#L83
Which I think is error prone in the current state anyway.

rullzer commented:
@nickvergessen ran the samshbox share tests and everything was OK.

rullzer commented:
So more comments/suggestions. Or shall we get this in!

DeepDiver1975 commented:
> So what we don't do now is in https://github.com/owncloud/core/blob/master/lib/private/share/helper.php#L83
> Which I think is error prone in the current state anyway.

But this is only necessary for reshares - right?

DeepDiver1975 commented:
@rullzer Oracle doesn't like your code :wink:

rullzer commented:
> @rullzer Oracle doesn't like your code :wink: 

That is fine. I'm not a big oracle fan myself :stuck_out_tongue_winking_eye: 
Let see

rullzer commented:
> But this is only necessary for reshares - right?

Yes

rullzer committed:
Fix oracle

DeepDiver1975 commented:
Haha - no db likes you now! :speak_no_evil:

nickvergessen commented on lib/private/share20/defaultshareprovider.php:
```diff
@@ -0,0 +1,257 @@
+		$qb->select('*')
+			->from('share')
+			->where($qb->expr()->eq('id', $qb->createParameter('id')))
+			->setParameter(':id', $id);
```
remove the leading `:` here to fix your tests?

rullzer commented:
Oracle is happy

schiesbn commented on lib/private/share20/defaultshareprovider.php:
```diff
@@ -0,0 +1,257 @@
+			'fileTarget' => $share->getTarget()
+		];
+
+		\OC_Hook::emit('OCP\Share', 'pre_unshare', $hookParams);
```
I think we should trigger the hooks by the manager and not by the provider. To make sure that the hooks are always triggered, no matter which share provider is used.

rullzer replied:
I think that is cleaner as well. But then we should also make sure that the deletion of children is actually handled from the manager. Else we would need callbacks from the providers to the manager and you have the same issue.

I agree. But lets do this in a followup PR.

schiesbn replied:
> I agree. But lets do this in a followup PR.

fine with me

nickvergessen replied:
Bwt we can pass objects in hooks? would it make more sense to send around the IShare?

rullzer replied:
I think we can. But lets do that in a later stage. Because then we would not to change all listners to unshare as well.

schiesbn commented:
beside the hooks which would move the the manager, as described here https://github.com/owncloud/core/pull/19872#discussion_r43764290 I would merge it so that we can move on. We can still adjust if needed during the next development steps :+1:

rullzer commented:
One more reviewer needed... lets get this in so sharing2.0 work can continue!

CC @DeepDiver1975 @nickvergessen

nickvergessen commented on apps/files_sharing/api/ocssharewrapper.php:
```diff
@@ -0,0 +1,51 @@
+	 * @return Share20OCS
+	 */
+	private function getShare20OCS() {
+		return new share20OCS(new \OC\Share20\Manager(
```
class name should be upper case (something for later)

nickvergessen commented on apps/files_sharing/api/ocssharewrapper.php:
```diff
@@ -0,0 +1,51 @@
+<?php
```
missing license (please add in follow up)

nickvergessen commented on apps/files_sharing/api/share20ocs.php:
```diff
@@ -0,0 +1,63 @@
+<?php
```
missing license (please add in follow up)

nickvergessen commented on apps/files_sharing/tests/api/share20ocstest.php:
```diff
@@ -0,0 +1,95 @@
+<?php
```
missing license (please add in follow up)

nickvergessen commented:
fine by me, remaining stuff can be fixed in follow up.

Especially the license needs to be done ;)

:+1:

rullzer commented:
@DeepDiver1975 Integration tests are happy :)
Only oracle is having backgroundjob failures again.
